### PR TITLE
[Taras Yakymiuk]Uit32 node server

### PR DESCRIPTION
**Pull request by taras1313**

### CODE REVIEW

File: src/server/controller.js [hunk 1]
```diff
@@ -1,15 +1,64 @@
...
     getProductById,
+    getHomepage,
+    addSubscription,
+    deleteSubscription,
     notFound,
 }
 
+
+function addSubscription(req, res) {
+    const emailPattern = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
```
Comment (dverbovyi) [inline src/server/controller.js]: in current implementation, regExp object will be instantiated  from scratch per every request.

move this pattern out of this function
```
const EMAIL_PATTERN = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
```

File: src/server/controller.js [hunk 2]
```diff
@@ -1,15 +1,64 @@
...
+    const randomProducts = new Set();
+    while (Array.from(randomProducts).length !== 6) {
+        const cleanedUpProduct = cleanUpProductProperties(productsMOCK[Math.floor(Math.random() * productsMOCK.length)])
+        randomProducts.add(cleanedUpProduct)
+    }
+
+    const homePageAggregated = {
+        newArrivals: Array.from(randomProducts),
+        slideShow: [
+            'https://myupdateweb.com/wp-content/uploads/2017/07/Bluehost.com_-1-1540x650.png',
```
Comment (dverbovyi) [inline src/server/controller.js]: put assets links here

File: src/server/controller.js [hunk 3]
```diff
@@ -1,15 +1,64 @@
...
+            'http://forgedground.com/image/cache/catalog/Slideshow/forged-ground-featured-official-merch-es-1540x650.jpg',
+            'http://forgedground.com/image/cache/catalog/viking-medieval-escudos-cascos-cuernos-espadass-accesorios-forgedground-1540x650.jpg',
+            'http://www.opencart.lionode.com/leoc04_2_2018/oc01/image/cache/catalog/banner%20main1-1540x650.jpg',
+            'http://www.opencart.lionode.com/leoc04_2_2018/oc01/image/cache/catalog/banner%20main2-1540x650.jpg'
+        ],
+        advArray: [{
```
Comment (dverbovyi) [inline src/server/controller.js]: wrong contract

screen shot 2019-02-15 at 11 22 48

File: src/server/controller.js [hunk 4]
```diff
@@ -59,7 +115,7 @@ function getProducts(req, res) {
 function getProductById(req, res) {
     const product = JSON.parse(JSON.stringify(productsMOCK.find((({ id }) => id === req.params.id))));
 
-    if(!product) notFound(req, res);
+    if (!product) notFound(req, res);
```
Comment (dverbovyi) [inline src/server/controller.js]: missed return

File: src/server/test.spec.js [hunk 5]
```diff
@@ -4,7 +4,7 @@ const request = require('supertest'),
     app = require('./app'),
     productsMocks = require('../assets/products.json');
 
-describe('#API TESTS', () => {
+describe('should return status 200, if get is successful', () => {
   describe('Accounts', () => {
```
Comment (dverbovyi) [inline src/server/test.spec.js]: accounts it not actual

**Review by dverbovyi — CHANGES_REQUESTED** [review]

File: src/server/router.js [hunk 6]
```diff
@@ -4,8 +4,12 @@ let express = require('express');
 let controller = require('./controller');
 let router = express.Router();
 
+// router.get('/api/homepage', memcacheMiddleWare, controller.getHomepage);
```
Comment (Roman-Klimkevych) [inline src/server/router.js]: Remove commented code, or add a TODO comment explaining why it is commented

Comment (taras1313) [reply src/server/router.js]: done

File: src/server/test.spec.js (hunk 5 above)
Comment (taras1313) [reply src/server/test.spec.js]: done

File: src/server/controller.js (hunk 4 above)
Comment (taras1313) [reply src/server/controller.js]: done

File: src/server/controller.js (hunk 3 above)
Comment (taras1313) [reply src/server/controller.js]: done

File: src/server/controller.js (hunk 2 above)
Comment (taras1313) [reply src/server/controller.js]: done

File: src/server/controller.js (hunk 1 above)
Comment (taras1313) [reply src/server/controller.js]: done

File: src/server/controller.js [hunk 7]
```diff
@@ -18,28 +39,86 @@ function cleanUpProductProperties(product) {
...
+function getHomepage(req, res) {
+    const randomProducts = new Set();
+    while (Array.from(randomProducts).length !== 6) {
+        const cleanedUpProduct = _cleanUpProductProperties(productsMOCK[Math.floor(Math.random() * productsMOCK.length)])
+        randomProducts.add(cleanedUpProduct)
+    }
+
+    const homePageAggregated = {
+        slideshow: [
+            'assets/img/slideshow/banner-main1-1540x650.jpg',
```
Comment (dverbovyi) [inline src/server/controller.js]: create slideshow.json, banners.json

File: src/server/router.js [hunk 8]
```diff
@@ -1,11 +1,16 @@
...
+const express = require('express');
+const controller = require('./controller');
+const memcacheMiddleware = require('./memcache.middleware');
+const router = express.Router();
 
+router.get('/api/homepage', controller.getHomepage);
+router.get('/api/filters', controller.getFilters);
 router.get('/api/products/:id', controller.getProductById);
-router.get('/api/products', controller.getProducts);
+router.get('/api/products', memcacheMiddleware(), controller.getProducts);
```
Comment (dverbovyi) [inline src/server/router.js:11]: apply middleware for all GET requests

File: src/server/test.spec.js [hunk 9]
```diff
@@ -4,8 +4,8 @@ const request = require('supertest'),
     app = require('./app'),
     productsMocks = require('../assets/products.json');
 
-describe('#API TESTS', () => {
-  describe('Accounts', () => {
+describe('should return status 200, if get is successful', () => {
+  describe('200', () => {
```
Comment (dverbovyi) [inline src/server/test.spec.js:8]: describe('Products', () => {...})

**Review by dverbovyi — CHANGES_REQUESTED** [review]

**Review by dverbovyi — APPROVED** [review]
@taras1313 merging now. Please create a separate PR for the following 
- fix the issue with middleware I had mentioned in direct messages (slack) 
- and add more tests
- update README with API docs

**Pull request merged**